### ISSUE TITLE
"lookup: lookup failed to find ..." error in template deployment outputs with modules

### ISSUE THREAD

**Issue by alastairtree** [terraform-providers/terraform-provider-azurerm#2836]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.11.9
provider.azurerm v1.21.0

### Affected Resource(s)

* `azurerm_template_deployment`

### Terraform Configuration Files

STAGE 1

> FILE: main.tf

```hcl
resource "azurerm_resource_group" "test" {
  name     = "delete-me"
  location = "North Europe"
}

module "dummy_module" {
  source         = "./dummy/"
  name           = "dummy-name"
  resource_group = "${azurerm_resource_group.test.name}"
}
```

>FILE: dummy/main.tf

```hcl
variable resource_group {}
variable name {}

resource "azurerm_template_deployment" "dummy" {
  name                = "sample-template"
  resource_group_name = var.resource_group

  template_body = <<EOF
{
    "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "name": {
            "type": "string"
        }
    },
    "variables": {},
    "resources": [ ]    
}
EOF

  parameters = {
    name = var.name
  }

  deployment_mode = "Incremental"
}
```

STAGE 2

> FILE: main.tf

```hcl
resource "azurerm_resource_group" "test" {
  name     = "delete-me"
  location = "North Europe"
}

module "dummy_module" {
  source         = "dummy/"
  name           = "dummy-name"
  resource_group = azurerm_resource_group.test.name
}

output "name" {
  value = module.dummy_module.name
}
```

>FILE: dummy/main.tf

```hcl
variable resource_group {}
variable name {}

resource "azurerm_template_deployment" "dummy" {
  name                = "sample-template"
  resource_group_name = var.resource_group

  template_body = <<EOF
{
    "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "name": {
            "type": "string"
        }
    },
    "variables": {},
    "resources": [  ]    
    ,"outputs": {
        "name": {
          "type": "string",
          "value": "[parameters('name')]"
        }
      }
}
EOF

  parameters = {
    name = var.name
  }

  deployment_mode = "Incremental"
}

output "name" {
  value = azurerm_template_deployment.dummy.outputs["name"]
}
```

### Expected Behavior

Deploy should succeed

### Actual Behavior

```
Error: Error refreshing state: 1 error(s) occurred:

* module.dummy_module.output.name: lookup: lookup failed to find 'name' in:

${lookup(azurerm_template_deployment.dummy.outputs, "name")}
```

### Steps to Reproduce

1. Setup tf files for stage 1 - with an azure resource group and a template deployment in a module with no outputs
1. `terraform init` and  `terraform apply`. Will deploy fine.
1. Edit tf files for stage 2 - i.e add an output to the template deployment, expose that output in the module and consume output in the parent.
1. `terraform apply`. Will error on state refresh ` lookup failed to find 'name' in:`

### Important Factoids

* The above is a simpified repro of a bug that seems to affect any use of output with template based deployments and modules, where you have already got the templated resource deployed in the state before you later decide to add any outputs.
* The bug can be worked around by removing the output from the tf file (but leaving it in the template json), running apply so that the output is available in state, and then adding the output again to the tf file and running apply again. This is hard to manage and awkward.
* This actually causes much more dev pain than you would think given the obscure looking repro above because:
  - we have to use templates because proper support of API management resources is lacking
  - we use modules to minimise template duplication and share common config
  - you often need to use outputs to force implicit dependencies between resources across modules with templates because modules do not support `depends_on` in the terraform resource graph

In conclusion, with substantial use of modules + templates andf following sensible principals of code reuse  you seem to end up hitting this bug which is pretty tough to diagnose and work around.

**Comment by alastairtree** [comment]
Any chance of an update from @tombuildsstuff or @katbyte ?

**Comment by TraGicCode** [comment]
Just ran into this issue as well.  :/ .

**Comment by ChethanRavulaGit** [comment]
I am facing the same issue without using modules, I am using azurerm_template_deployment to run ARM template to create Web App. I have outputs defined for Principal ID and Tenant Id as below
"outputs": {
      "principal_Id": {
        "type": "string",
        "value": "[reference(concat('Microsoft.Web/sites/', parameters('WebApp_Name')), '2018-02-01', 'Full').identity.principalId]"
      },
      "tenant_Id": {
        "type": "string",
        "value": "[reference(concat('Microsoft.Web/sites/', parameters('WebApp_Name')), '2018-02-01', 'Full').identity.tenantId]"
      }
    }

When i run terraform plan, i am seeing the below error.
lookup: lookup failed to find 'tenant_Id' in:
lookup: lookup failed to find 'principal_Id' in:

Anybody able to get outputs from ARM successfully?

**Comment by tombuildsstuff** [comment]
@ChethanRavulaGit I'm not sure if you've seen but the `azurerm_app_service` resource natively supports `identity` - so you should be able to use this instead of the `azurerm_template_deployment` resource: https://www.terraform.io/docs/providers/azurerm/r/app_service.html

**Comment by ChethanRavulaGit** [comment]
@tombuildsstuff ; thanks for the response but I am not able to provision web app within ASE using terraform. When i use Terraform for creating Web App within ASE, it takes long time and errors out.
So i am forced to use ARM for this.

Also; i was able to resolve the issue with outputs. It worked fine when i deleted the Statefiles and recreated them

**Comment by rpiteira** [comment]
Just run on this problem yesterday! Any update on this?

**Comment by ChethanRavulaGit** [comment]
@rpiteira 
i was not able to solve the issue using Terraform but could work around using ARM and calling it from Terraform.

**Comment by alastairtree** [comment]
Have just retested this with latest terraform and azurerm and still broken. Have updated the OP to use modern terraform syntax and no longer using lookup function. The new error message is:
```
Error: Invalid index

  on dummy\main.tf line 36, in output "name":
  36:   value = azurerm_template_deployment.dummy.outputs["name"]
    |----------------
    | azurerm_template_deployment.dummy.outputs is empty map of string

The given key does not identify an element in this collection value.
```

However there is an easy workaround - add the output to the template first and run apply. Then add the output to the module/main and apply again and it works. Given that I am gonna close it.
